Nos/Service/NamesAPI.swift

```diff
@@ -0,0 +1,110 @@
+//
+//  NamesAPI.swift
+//  Nos
+//
+//  Created by Martin Dutra on 21/2/24.
+//
+
+import Foundation
+
+class NamesAPI {
+
+    private enum Error: LocalizedError {
+        case unexpected
+        case usernameNotAvailable
+
+        var errorDescription: String? {
+            switch self {
+            case .unexpected:
+                return "Unexpected"
+            case .usernameNotAvailable:
+                return String(localized: .localizable.usernameAlreadyClaimed)
+            }
+        }
+    }
+
+    private enum HTTPMethod: String {
+        case delete = "DELETE"
+        case post = "POST"
+    }
+
+    private let verificationURL: URL
+    private let registrationURL: URL
+
+    init?(host: String = "nos.social") {
+        guard let verificationURL = URL(string: "https://\(host)/.well-known/nostr.json") else {
+            return nil
+        }
+        guard let registrationURL = URL(string: "https://\(host)/api/names") else {
+            return nil
+        }
+        self.verificationURL = verificationURL
+        self.registrationURL = registrationURL
+    }
+
+    func delete(username: String, keyPair: KeyPair) async throws {
+        let request = try buildURLRequest(
+            url: registrationURL.appending(path: username),
+            method: .delete,
+            json: try JSONSerialization.data(
+                withJSONObject: ["name": username, "data": ["pubkey": keyPair.publicKeyHex]]
+            ),
+            keyPair: keyPair
+        )
+        let (_, response) = try await URLSession.shared.data(for: request)
+        if let response = response as? HTTPURLResponse {
+            if response.statusCode == 200 {
+                return
+            }
+        }
+        throw Error.unexpected
+    }
+
+    func verify(username: String) async throws -> Bool {
+        let request = URLRequest(
+            url: verificationURL.appending(queryItems: [URLQueryItem(name: "name", value: username)])
+        )
+        let (_, response) = try await URLSession.shared.data(for: request)
+        if let response = response as? HTTPURLResponse {
+            return response.statusCode == 404
+        } else {
+            return false
+        }
+    }
+
+    func register(username: String, keyPair: KeyPair) async throws {
+        let request = try buildURLRequest(
+            url: registrationURL,
+            method: .post,
+            json: try JSONSerialization.data(
+                withJSONObject: ["name": username, "data": ["pubkey": keyPair.publicKeyHex]]
+            ),
```

Comment:
This is looking much better with the `buildURLRequest` method. It looks like both register and delete create exactly the same JSON here, right? Could we move this into the `buildURLRequest` method?

I might also remove the `url` parameter and just create the URL inside of `buildURLRequest`, too. I see it's different whether you're registering or deleting but we could replace the HTTPMethod parameter with an enum that's specific to the NamesAPI, like NamesRequestType, maybe:

```
enum NamesRequestType {
    case delete
    case register
}
```

Or maybe even taking it further:

```
enum NamesRequest {
    case delete(username: String, keyPair: KeyPair)
    case register(username: String, keyPair: KeyPair)
    case verify(username: String)

    var httpMethod: HTTPMethod { ... }
    var jsonEvent: JSONEvent? { ... }
    var url: URL { ... }
    var urlRequest: URLRequest { ... } // the buildURLRequest method could live here
}
```

This might be taking things a little too far, but I'd love if we could be consistent and concise between all three methods (`delete`, `verify`, and `register`) so that's what I'm attempting with these ideas.

Reply:
It's true the HTTP body are the same, I didn't notice it. I refactored it to a distinct function that both can use so it is clear the body is the same and we avoid repetition.

I'm not convinced about the other suggestion though, not only we would be adding an additional responsibility to buildURLRequest (knowing which URL to use in each operation) but we could be over engineering on the basis that two of the three operations we have are very similar, but we could need to revert it if we add another operation that breaks the rule.

IMO, what we have is a buildURLRequest that handles the authentication part (which is shared by all operations) and a build JSON that handles the body creation in the two operations we now have and happen to share the same structure,

Reply:
Makes sense. I'm fine with leaving this as-is.